File: framework/languages/java/samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/actors/SupportUserActorTransferAdapter.java
package systems.zlink.samples.supportchat.server.support.actors;

import systems.zlink.framework.actors.ZLinkActorContext;
import systems.zlink.framework.actors.ZLinkActorTransferAdapter;
import systems.zlink.framework.messaging.ZLinkMessage;

public final class SupportUserActorTransferAdapter
    implements ZLinkActorTransferAdapter<SupportUserActor> {
    @Override
    public java.util.concurrent.CompletionStage<ZLinkMessage> transferOut(SupportUserActor actor) {
        return java.util.concurrent.CompletableFuture.completedFuture(
            ZLinkMessage.of(new TransferState(
                actor.displayName(), actor.role(), actor.participantId(), actor.conversationId())));
    }

    @Override
    public java.util.concurrent.CompletionStage<SupportUserActor> transferIn(
        String actorId,
        ZLinkActorContext context,
        ZLinkMessage state) {
        TransferState transferred = state.decode(TransferState.class);
        SupportUserActor actor = new SupportUserActor(actorId, context);
        actor.setIdentity(transferred.displayName(), transferred.role(), transferred.participantId());
        actor.joinConversation(transferred.conversationId());
        return java.util.concurrent.CompletableFuture.completedFuture(actor);
    }

    public record TransferState(
        String displayName,
        String role,
        String participantId,
        String conversationId) {
    }
}

File: framework/languages/java/samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/actors/SupportUserActorTransferAdapter.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors

import systems.zlink.framework.actors.ZLinkActorContext
import systems.zlink.framework.kotlin.ZLinkSuspendingActorTransferAdapter
import systems.zlink.framework.messaging.ZLinkMessage

class SupportUserActorTransferAdapter : ZLinkSuspendingActorTransferAdapter<SupportUserActor>() {
    override suspend fun transferOutSuspending(
        actor: SupportUserActor,
    ): ZLinkMessage = ZLinkMessage.of(
        TransferState(actor.displayName, actor.role, actor.participantId, actor.conversationId),
    )

    override suspend fun transferInSuspending(
        actorId: String,
        context: ZLinkActorContext,
        state: ZLinkMessage,
    ): SupportUserActor {
        val transferred = state.decode(TransferState::class.java)
        return SupportUserActor(actorId, context).also { actor ->
            actor.setIdentity(transferred.displayName, transferred.role, transferred.participantId)
            if (transferred.conversationId.isNotBlank()) {
                actor.joinConversation(transferred.conversationId)
            }
        }
    }

    data class TransferState(
        val displayName: String,
        val role: String,
        val participantId: String,
        val conversationId: String,
    )
}
